realm/realm-library/src/main/java/io/realm/RealmCache.java

```diff
@@ -85,45 +91,66 @@ static RealmCacheType valueOf(Class<? extends BaseRealm> clazz) {
-        configuration = config;
+    private RealmCache() {
         refAndCountMap = new EnumMap<>(RealmCacheType.class);
         for (RealmCacheType type : RealmCacheType.values()) {
             refAndCountMap.put(type, new RefAndCount());
         }
     }
 
+    private static RealmCache getCache(String realmPath, boolean createIfNotExist) {
+        RealmCache cacheToReturn = null;
+        synchronized (cachesMap) {
+            Iterator<Map.Entry<String, WeakReference<RealmCache>>> it = cachesMap.entrySet().iterator();
+            while (it.hasNext()) {
+                Map.Entry<String, WeakReference<RealmCache>> entry = it.next();
+                RealmCache cache = entry.getValue().get();
+                if (cache == null) {
+                    // Clear the entry if there is no one holding the RealmCache.
+                    it.remove();
+                } else if (entry.getKey().equals(realmPath)) {
+                    cacheToReturn = cache;
+                }
+            }
+
+            if (cacheToReturn == null && createIfNotExist) {
+                cacheToReturn = new RealmCache();
+                cachesMap.put(realmPath, new WeakReference<RealmCache>(cacheToReturn));
+            }
+        }
+        return cacheToReturn;
+    }
+
     /**
      * Creates a new Realm instance or get an existing instance for current thread.
      *
      * @param configuration {@link RealmConfiguration} will be used to create or get the instance.
      * @param realmClass class of {@link Realm} or {@link DynamicRealm} to be created in or gotten from the cache.
      * @return the {@link Realm} or {@link DynamicRealm} instance.
      */
-    static synchronized <E extends BaseRealm> E createRealmOrGetFromCache(RealmConfiguration configuration,
+    static <E extends BaseRealm> E createRealmOrGetFromCache(RealmConfiguration configuration,
             Class<E> realmClass) {
-        boolean isCacheInMap = true;
-        RealmCache cache = cachesMap.get(configuration.getPath());
-        if (cache == null) {
-            // Creates a new cache.
-            cache = new RealmCache(configuration);
-            // The new cache should be added to the map later.
-            isCacheInMap = false;
+        RealmCache cache = getCache(configuration.getPath(), true);
 
-            copyAssetFileIfNeeded(configuration);
-        } else {
-            // Throws the exception if validation failed.
-            cache.validateConfiguration(configuration);
-        }
+        return cache.doCreateRealmOrGetFromCache(configuration, realmClass);
+    }
+
+    private synchronized <E extends BaseRealm> E doCreateRealmOrGetFromCache(RealmConfiguration configuration,
+            Class<E> realmClass) {
 
-        RefAndCount refAndCount = cache.refAndCountMap.get(RealmCacheType.valueOf(realmClass));
+        RefAndCount refAndCount = refAndCountMap.get(RealmCacheType.valueOf(realmClass));
+
+        if (getTotalGlobalRefCount() == 0) {
```

Comment:
I think `globalRefCount()` could be 0 if `Realm realm = Realm.getDefaultInstance(); realm.close();`